### PR TITLE
Move the Cart Store dependency out of build chunk: try 2

### PR DESCRIPTION
**Pull request by jsnajdr**
Second attempt at landing #26303.

**Reorg the `lib/domains/utils` module**
Move the function `getDomainNameFromReceiptOrCart` that has some heavy dependencies into a separate module. Then Webpack will not bundle the whole `lib/domains/utils` tree into `build`, but will be able to split the modules between chunks better. Two separate modules can be bundled into chunks separately, each one in different chunk.

**Invert the control flow when Cart Store listens for selected site changes in Redux**
There is a `state/lib/middleware.js` module that imports the Cart Store and call a `setSelectedSiteId` method on it on every Redux selected site update. That means that Cart Store gets bundled into `build`.

This patch inverts the control flow: when Cart Store is first loaded (by someone who **really** needs it), it subscribes to Redux updates (using the Redux Bridge) and updates itself when needed. That removes the Cart Store from the `build` chunk.

We should eventually use this technique for all subscribers in `state/lib/middleware.js` and probably at many other places, too.

**How to test:**
This is the crucial flow that was broken in the previous attempt and is now fixed:
**1. Go to `/me/purchases/:siteId/:purchaseId`**
This is a page that shows particular purchase for a particular site. The site is not necessarily an active site present in `state.sites`. For example, disconnected Jetpack sites are no longer active and yet can have a valid purchase we want to see. Therefore, we can't just set `state.ui.selectedSiteId` and let `CartStore` pick up the change. The `CartStore`'s site ID must be set by directly calling `CartStore.setSelectedSiteId`.

**2. Click "Renew Now" (the viewed purchase must be renewable)**
This will do two things:
- add a new product to `CartStore`. Because the cart hasn't been loaded yet (it's loaded only when poller issues `CartSynchronizer._poll` request and that happens only after `CartStore` gets its first `change` listener), the change is enqueued.
- navigate to `/checkout/:siteId`

It's absolutely crucial that the navigation to `/checkout/:siteId` doesn't change the `CartStore`'s selected site ID. If it did change the ID, `CartStore` would throw away the enqueued changes and load a new cart for a new site. The cart would end up being empty and in that case the `/checkout/:siteId` route [redirects](https://github.com/Automattic/wp-calypso/blob/7f1ba4f2deea6dc83019fe26c4fb6c02298c3a38/client/my-sites/checkout/checkout/index.jsx#L287-L321) to `/plans`.

**3. Verify that you see a checkout page with the right products in the cart.**

### CODE REVIEW

**Comment by alisterscott** [comment]
Is this still being worked on?

**Comment by jsnajdr** [comment]
Yep, still worked on.

**Comment by jsnajdr** [comment]
This refactoring that removes a substantial portion of the initial `build` chunk is finally, after several months, ready for another review round.

The PR description contains instructions how to test the "Renew Now" flow, the one that broke on the previous attempt (#26303, Aug 2018)

**Review by blowery — APPROVED** [review]
LGTM!

**Comment by jsnajdr** [comment]
Thanks for the approval @blowery 👍 I'll wait for @klimeryk and team to give this PR a second review.

**Review by klimeryk — CHANGES_REQUESTED** [review]
Thanks so much for pushing for this refactor!

Tested all the cases I could think of:

 - regular domain purchase in Calypso and signup
 - having multiple domains in cart
 - having something in cart and then trying to renew an existing purchase -> cart only has the renewal item, as expected
 - renewing a domain that's in redemption
 - buying a domain from `/domains` using the siteless checkout (`no-site`)
 - pre-populating the shopping cart based on the renewal link from the email reminders we send

All looked super solid... but I managed to break it at the very last minute as I was writing up all these findings 😅 

Basically:

 - have a domain name that's expiring (let me know if you need an example)
 - go to Manage -> Domains -> click on the domain and then the `Renew it now.` link (it looks like `/checkout/domain_map:example.com/renew/6163153/example.wordpress.com`
 - you'll go successfully to checkout, with the correct item in the cart
 - however, if you try to _remove_ the item from the cart clicking the 🗑 icon, the screen will go blank and you'll get this error in the console:

```
Uncaught Error: siteName and purchaseId must be provided
    at managePurchase (paths.js:27)
    at getExitCheckoutUrl (index.js:30)
    at Checkout.redirectIfEmptyCart (index.jsx:563)
    at Checkout.componentDidUpdate (index.jsx:402)
    at commitLifeCycles (react-dom.development.js:17349)
    at commitAllLifeCycles (react-dom.development.js:18736)
    at HTMLUnknownElement.callCallback (react-dom.development.js:149)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:199)
    at invokeGuardedCallback (react-dom.development.js:256)
    at commitRoot (react-dom.development.js:18948)
```

Meaning, this line is triggered: https://github.com/Automattic/wp-calypso/blob/45c1018c8a5f10682b75a06ce799c9d6f75aea3a/client/me/purchases/paths.js#L20

Call stack:
```
managePurchase (paths.js:26)
getExitCheckoutUrl (index.js:30)
redirectIfEmptyCart (index.jsx:563)
componentDidUpdate (index.jsx:402)
commitLifeCycles (react-dom.development.js:17349)
commitAllLifeCycles (react-dom.development.js:18736)
callCallback (react-dom.development.js:149)
invokeGuardedCallbackDev (react-dom.development.js:199)
invokeGuardedCallback (react-dom.development.js:256)
commitRoot (react-dom.development.js:18948)
(anonymous) (react-dom.development.js:20418)
unstable_runWithPriority (scheduler.development.js:255)
completeRoot (react-dom.development.js:20417)
performWorkOnRoot (react-dom.development.js:20346)
performWork (react-dom.development.js:20254)
performSyncWork (react-dom.development.js:20228)
interactiveUpdates$1 (react-dom.development.js:20495)
interactiveUpdates (react-dom.development.js:2170)
dispatchInteractiveEvent (react-dom.development.js:4882)
```

Inspecting the variables in the debugger, looks like the `purchaseId` is `undefined`.

Tested it on wpcalypso (since this check is skipped on production) and could not reproduce it, so I think it might be related to the changes here.
Naturally, it does not seem like a very common flow (though, I can see how some users could try to renew a domain and then "cancel" it like that once they see the price), so certainly not impossible.
It won't blank the screen on production and probably a safeguard will kick in that will simply redirect the user to `/me/purchases`, but it seems worth investigating in case there are other flows that could trigger this state.

**Review by klimeryk — APPROVED** [review]
Huh, you're right - there's the undefined redirect on wpcalypso 🤔 I was pretty sure it was an existing behaviour, but got confused when I did not get the error on `wpcalypso` (which is not production, right?). But, yeah, looks unrelated, so :shipit: !

**Pull request merged**
